src/pomelo_plug.c

```diff
@@ -15,11 +15,144 @@
+int PHS_pomelo(void *out, size_t outlen, const void *in, size_t inlen, const void *salt, size_t saltlen, unsigned int t_cost, unsigned int m_cost);
+
+int PHS_pomelo(void *out, size_t outlen, const void *in, size_t inlen, const void *salt, size_t saltlen, unsigned int t_cost, unsigned int m_cost)
+{
+    unsigned long long i, j;
+    unsigned long long i0, i1, i2, i3, i4;
+    __m256i *S;
+    unsigned long long random_number, index_global, index_local;
+    unsigned long long state_size, mask;
 
-#ifdef __SSE2__
+    //check the size of password, salt, and output. Password at most 256 bytes; salt at most 64 bytes; output at most 256 bytes.
+    if (inlen > 256 || saltlen > 64 || outlen > 256 || inlen < 0 || saltlen < 0 || outlen < 0) return 1;
+
+    //Step 1: Initialize the state S
+    state_size = 1ULL << (13+m_cost);    // state size is 2**(13+m_cost) bytes
+    S = (__m256i *)mem_alloc_align(state_size, 32);   // aligned malloc is needed; otherwise it is only aligned to 16 bytes when using GCC.
+    mask = (1ULL << (8+m_cost)) - 1;     // mask is used for modulation: modulo size_size/32
+
+    //Step 2: Load the password, salt, input/output sizes into the state S
+    for (i = 0; i < inlen; i++)   ((unsigned char*)S)[i] = ((unsigned char*)in)[i];         // load password into S
+    for (i = 0; i < saltlen; i++) ((unsigned char*)S)[inlen+i] = ((unsigned char*)salt)[i]; // load salt into S
+    for (i = inlen+saltlen; i < 384; i++) ((unsigned char*)S)[i] = 0;
+    ((unsigned char*)S)[384] = inlen & 0xff;          // load password length (in bytes) into S;
+    ((unsigned char*)S)[385] = (inlen >> 8) & 0xff;   // load password length (in bytes) into S;
+    ((unsigned char*)S)[386] = saltlen;               // load salt length (in bytes) into S;
+    ((unsigned char*)S)[387] = outlen & 0xff;         // load output length (in bytes into S)
+    ((unsigned char*)S)[388] = (outlen >> 8) & 0xff;  // load output length (in bytes into S)
+    ((unsigned char*)S)[389] = 0;
+    ((unsigned char*)S)[390] = 0;
+    ((unsigned char*)S)[391] = 0;
+
+    ((unsigned char*)S)[392] = 1;
+    ((unsigned char*)S)[393] = 1;
+    for (i = 394; i < 416; i++) ((unsigned char*)S)[i] = ((unsigned char*)S)[i-1] + ((unsigned char*)S)[i-2];
+
+    //Step 3: Expand the data into the whole state
+    for (i = 13; i < (1ULL << (8+m_cost)); i=i+1)  F0(i);
+
+    //Step 4: Update the state using function G
+    random_number = 123456789ULL;
+    for (i = 0; i < (1ULL << (7+m_cost+t_cost)); i=i+32)  G(i,random_number);
+
+    //Step 5: Update the state using function H
+    for (i = 1ULL << (7+m_cost+t_cost);  i < (1ULL << (8+m_cost+t_cost)); i=i+32)  H(i,random_number);
+
+    //Step 6: Update the state using function F
+    for (i = 0; i < (1ULL << (8+m_cost)); i=i+1)  F(i);
+
+    //Step 7: Generate the output
+    memcpy(out, ((unsigned char*)S)+state_size-outlen, outlen);
+    memset(S, 0, state_size);  // clear the memory
+    MEM_FREE(S);           // free the memory
```

Comment:
This reminds us that POMELO is one of those large-memory hashes where we'd want to keep the memory (de)allocation out of the loop, reusing (y)escrypt's per-thread local memory regions concept & code. Indeed, this is probably not worth the time for a format that won't actually be used, but it may be good for practice. ;-)